src/Protocols/Df.hs

```diff
@@ -635,64 +635,84 @@ registerBwd
-    -- initialize bram
-    brRead = C.readNew
-             (C.blockRamU C.NoClearOnReset fifoDepth (C.errorX "No reset function"))
-             brReadAddr brWrite
-    -- run the state machine (a mealy machine)
-    (brReadAddr, brWrite, otpA, otpB)
-      = C.unbundle
-      $ C.mealy machineAsFunction s0
-      $ C.bundle (brRead, C.unsafeToHighPolarity reset, inpA, inpB)
-
-  -- when reset is on, set state to initial state and output blank outputs
-  machineAsFunction _ (_, True, _, _) = (s0, (0, Nothing, Ack False, NoData))
-  machineAsFunction (rAddr0,wAddr0,amtLeft0) (brRead0, False, pushData, Ack popped) =
-    let -- potentially push an item onto blockram
-        maybePush = if amtLeft0 > 0 then dataToMaybe pushData else Nothing
-        brWrite = (wAddr0,) <$> maybePush
-        -- adjust write address and amount left
-        --   (output state machine doesn't see amountLeft')
-        (wAddr1, amtLeft1)
-          | Just _ <- maybePush = (C.satSucc C.SatWrap wAddr0, amtLeft0 - 1)
-          | otherwise           = (wAddr0, amtLeft0)
-        -- if we're about to push onto an empty queue, we can pop immediately instead
-        (brRead1, amtLeft2)
-          | Just push <- maybePush, amtLeft0 == maxBound = (push, amtLeft1)
-          | otherwise                                    = (brRead0, amtLeft0)
-        -- adjust blockram read address and amount left
-        (rAddr1, amtLeft3)
-          | amtLeft2 < maxBound && popped = (C.satSucc C.SatWrap rAddr0, amtLeft1 + 1)
-          | otherwise                     = (rAddr0, amtLeft1)
-        brReadAddr = rAddr1
-        -- return our new state and outputs
-        otpAck = Maybe.isJust maybePush
-        otpDat = if amtLeft2 < maxBound then Data brRead1 else NoData
-    in  ((rAddr1, wAddr1, amtLeft3), (brReadAddr, brWrite, Ack otpAck, otpDat))
-
-  -- initial state
-  -- (next read address in bram, next write address in bram, space left in bram)
-  -- Addresses only go from 0 to depth-1.
-  -- Space left goes from 0 to depth because the fifo could be empty
-  -- (space left = depth) or full (space left = 0).
-  s0 :: (C.Index depth, C.Index depth, C.Index (depth C.+ 1))
-  s0 = (0, 0, maxBound)
+  -- | Consumes `Df dom a`, produces `Df dom a` along with ready signal and data count.
+    Circuit (Df dom a) (Df dom a)
+fifo depth = Circuit go
+ where
+  go (m2sL, s2mR) = (s2mL, m2sR)
+   where
+    (s2mL, (m2sR, _, _)) = toSignals (fifoWithMeta depth) (m2sL, (s2mR, CSignal (C.pure ()), CSignal (C.pure ())))
+
+-- | State record for the FIFO circuit.
+data FifoState depth = FifoState
+  { readPointer   :: C.Index depth
+  , dataCount     :: C.Index (depth C.+ 1)
+  } deriving (Generic, C.NFDataX)
+
+-- | A generic First-In-First-Out (FIFO) circuit with a specified depth that exposes meta information such as a ready signal and data count.
+-- At least one cycle latency.
+-- When the reset is high or the enable is low, the output will be `Nothing` and the ready signal will be `False`.
```

Comment:
Please keep to 80 chars soft / 90 chars hard. Feel free to add

```json
  "editor.rulers": [
    80,
    90
  ],
```

to `.vscode/settings.json`.